VideoLocker/src/main/java/org/edx/mobile/view/DownloadListActivity.java

```diff
@@ -56,68 +38,57 @@ public void onItemClicked(DownloadEntry model) {
             }
 
             @Override
-            public void onDownloadComplete(DownloadEntry model) {
-                //showDownloadCompleteView();
-                adapter.remove(model);
-                adapter.notifyDataSetChanged();
-            }
-
-            @Override
             public void onDeleteClicked(DownloadEntry model) {
-                if(environment.getStorage().removeDownload(model) >= 1){
+                if (environment.getStorage().removeDownload(model) >= 1) {
                     // update the list data as one download is removed
-                    try{
-                        adapter.remove(model);
-                        adapter.notifyDataSetChanged();
-                    }catch(Exception e){
-                        logger.error(e);
-                    }
-
+                    adapter.remove(model);
+                    adapter.notifyDataSetChanged();
                 }
             }
         };
-
-        adapter.setStore(environment.getStorage());
+        final ListView downloadListView = (ListView) findViewById(R.id.my_downloads_list);
         downloadListView.setAdapter(adapter);
-        final ArrayList<DownloadEntry> list = new ArrayList<DownloadEntry>();
+    }
+
+    private void observeOngoingDownloads() {
         environment.getDatabase().getListOfOngoingDownloads(new DataCallback<List<VideoModel>>() {
             @Override
-            public void onResult(List<VideoModel> result) {
-                if(result!=null){
-                    for(VideoModel de : result){
-                        list.add((DownloadEntry) de);
-                    }
-                    adapter.setItems(list);
+            public void onResult(final List<VideoModel> result) {
+                if (result != null) {
+                    handler.post(new Runnable() {
+                        @Override
+                        public void run() {
+                            adapter.setItems((List) result);
+                        }
+                    });
                 }
             }
+
             @Override
             public void onFail(Exception ex) {
                 logger.error(ex);
             }
         });
+        handler.postDelayed(new Runnable() {
+            @Override
+            public void run() {
+                observeOngoingDownloads();
```

Comment:
Hmm, using a fixed-interval refresh loses the ordering guarantee of the previous sequential approach, though it's probably unlikely to be an issue in practice. I guess this was done in order to have an easy mechanism to stop the refresh cycle by just removing the pending callbacks from the `Handler`. This could potentially be addressed by keeping a reference to the callback for the last query and performing a check against it, or going back to the sequential approach and just checking for a 'resumed` in the callback. It's not likely to be an issue in practice though, so up to you if you want to address it.